rfcs/2021-02-04-5802-event-metadata.md

```diff
@@ -0,0 +1,276 @@
+is a data structure that can inline a (fixed) number of elements before
+allocating memory. By using this feature, this common case can avoid an
+extra allocation and improve data locality.
+
+Similarly, the introduction of
+[`SmolStr`](https://github.com/rust-analyzer/smol_str) is driven by the
+fact that source name and event IDs are highly likely to be short, will
+never be modified. Also, in the case of the source name, it will be
+shared with the actual source and all other events originating from that
+source, and `SmolStr` provides constant-time copy.
+
+## Drawbacks
+
+- The event data is a critical component of Vector. Any changes to this
+  structure impacts virtually the entire system.
+- All transforms that combine or duplicate events will have extra
+  complications in handing the metadata.
+- The user scripting transforms, Lua and WASM, will require extra work
+  from script writers to handle the metadata.
+- This metadata will increase the memory required by Vector for any
+  given event flow.
+- Recording, copying, and finalizing the metadata will increase the CPU
+  usage.
+
+## Alternatives
+
+The following alternative components have been considered and discarded:
+
+1. Reduce the event metadata to a UUID, and track the actual metadata in
+   a separate map (or set of maps, as appropriate). This has the
+   downside of requiring most of the work necessary to just attach the
+   metadata itself to the event, as well as introducing both the
+   overhead of map lookups and the possibility of causing memory leaks
+   due to map entries not being purged, without any real benefit.
+2. Use a dictionary (ie `HashMap`) of arbitrary values to hold the
+   metadata. While some of the metadata may end up being optional, this
+   adds overhead to all accesses, due to the hashing and conversion of
+   the resulting data, as well as more work for the allocator.
+3. Store each event source in an `Arc`. Since the event source is
+   read-only, there is no need for an additional `Mutex` layer, and this
+   can provide an easy performance boost when events are duplicated
+   between multiple transforms or sinks. However, this adds the overhead
+   of an extra allocation and reference counting for the `Arc`
+   box. Since common usage patterns do not duplicate events, the costs
+   outweigh the benefits on this one.
+
+## Plan Of Attack
+
+Incremental steps that execute this change. Generally this is in the form of:
+
+- [ ]  Create the new structures with an empty stub `EventMetadata`.
+- [ ]  Add a simple timestamp to the metadata to ensure it is propagated
+       from end to end.
+- [ ]  Add instrumentation to full `Event` creation and cloning (gated
+       by a feature flag) and benchmark the result to estimate the
+       performance impact of clone vs reference counting the metadata.
+- [ ]  Rework sources that use `Utc::now()` or equivalent to set a
+       default timestamp to use this new metadata (optimization,
+       optional).
```

Comment:
As mentioned above, we may want to expand a bit on this in terms of backwards compatibility.

Reply:
What I am describing here is purely an optimization for things that just use `Utc::now` as a default timestamp. I may not actually do this pass, but put it in because it seemed like a few uses might shake out as I worked on other things.